crates/sui-core/src/execution_driver.rs

```diff
@@ -83,6 +84,10 @@ pub async fn execution_process(
                     .try_execute_immediately(&certificate, &epoch_store)
                     .await;
                 if let Err(e) = res {
+                    if matches!(e, SuiError::WrongEpoch { .. }) {
+                        warn!("Scheduled outdated certificate {:?} for wrong epoch {:?}, not retrying", certificate, e);
+                        break;
+                    }
```

Comment:
@mwtian @williampsmith are there any places that do a `notify_read_effects` that might hang forever in this case?

Reply:
I *believe* this is fine... I don't think this should ever happen for checkpoint execution as it would panic if we're ever executing a checkpoint from the wrong epoch. For transaction orchestrator, it would timeout so also fine.

Reply:
1. Can we compare the epoch of the cert and epoch store before execution,  instead of detecting errors from execution?
2. For `notify_read_effects()` in RPC handlers, the server side timeout would terminate them eventually.

I guess if the logic to update authority's epoch store is correct, i.e. epoch store only gets updated when there is no more transaction that needs to be executed from the current epoch, this would be correct.